libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,905 @@
+ * @self: A #FuProgress
+ *
+ * Monitor a child and proxy back up to the parent with the correct percentage.
+ *
+ * Return value: (transfer none): A new %FuProgress or %NULL for failure
+ *
+ * Since: 1.7.0
+ **/
+FuProgress *
+fu_progress_get_child(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_PROGRESS(self), NULL);
+	g_return_val_if_fail(priv->id != NULL, NULL);
+
+	/* already created child */
+	if (priv->child != NULL)
+		return priv->child;
+
+	/* connect up signals */
+	priv->child = fu_progress_new(NULL);
+	priv->percentage_child_id =
+	    g_signal_connect(priv->child,
+			     "percentage-changed",
+			     G_CALLBACK(fu_progress_child_percentage_changed_cb),
+			     self);
+	priv->status_child_id = g_signal_connect(priv->child,
+						 "status-changed",
+						 G_CALLBACK(fu_progress_child_status_changed_cb),
+						 self);
+	fu_progress_set_parent(priv->child, self);
+	return priv->child;
+}
+
+static void
+fu_progress_show_profile(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	gdouble division;
+	gdouble total_time = 0.0f;
+	gboolean close_enough = TRUE;
+	g_autoptr(GString) str = NULL;
+
+	/* not accurate enough for a profile result */
+	if (priv->flags & FU_PROGRESS_FLAG_NO_PROFILE)
+		return;
+
+	/* get the total time so we can work out the divisor */
+	str = g_string_new("raw timing data was { ");
+	for (guint i = 0; i < priv->step_max; i++) {
+		FuProgressStep *step = g_ptr_array_index(priv->steps, i);
+		g_string_append_printf(str, "%.3f, ", step->profile);
+	}
+	if (priv->step_max > 0)
+		g_string_set_size(str, str->len - 2);
+	g_string_append(str, " } -- ");
+
+	/* get the total time so we can work out the divisor */
+	for (guint i = 0; i < priv->step_max; i++) {
+		FuProgressStep *step = g_ptr_array_index(priv->steps, i);
+		total_time += step->profile;
+	}
+	if (total_time < 0.001)
+		return;
+	division = total_time / 100.0f;
+
+	/* what we set */
+	g_string_append(str, "steps were set as [ ");
+	for (guint i = 0; i < priv->step_max; i++) {
+		FuProgressStep *step = g_ptr_array_index(priv->steps, i);
+		g_string_append_printf(str, "%u ", step->value);
+	}
+
+	/* what we _should_ have set */
+	g_string_append_printf(str, "] but should have been [ ");
+	for (guint i = 0; i < priv->step_max; i++) {
+		FuProgressStep *step = g_ptr_array_index(priv->steps, i);
+		g_string_append_printf(str, "%.0f ", step->profile / division);
+
+		/* this is sufficiently different to what we guessed */
+		if (fabs((gdouble)step->value - step->profile / division) > 5)
+			close_enough = FALSE;
+	}
+	g_string_append(str, "]");
+	if (!close_enough) {
+#ifdef SUPPORTED_BUILD
+		g_debug("%s at %s", str->str, priv->id);
+#else
+		if (priv->flags & FU_PROGRESS_FLAG_GUESSED) {
+			g_warning("%s at %s", str->str, priv->id);
+			g_warning("If not already fixed on master (please check!), please report "
+				  "this warning to us as a GitHub issue. Thanks!");
```

Comment:
I think putting this into a wiki page and better explaining what this is and what we want in the issue report will be useful.  Maybe even a github bug template for it.

Reply:
Fixed.